Annotate for trim/AOT (in)compatibility

sbomer opened the pull request:
Adds annotations to propagate trim and AOT compatibility warnings to callers. This way, a trimmed/AOT-compiled app referencing Newtonsoft.Json will get warnings at the places where it calls into Newtonsoft.Json, instead of warnings that point at Newtonsoft.Json internals.

Adds a net8.0 target, because the AOT analyzer requires net7.0+.

I'm not setting up a test app because I don't want to be responsible for some ad-hoc infrastructure or script that does the validation - but I did check this change against a test app locally. I think relying on the analyzer is good enough in this case because the goal isn't to make the code trim-compatible, but to surface existing trim incompatibilities earlier in the development cycle for consumers. If we do miss some warnings it's not a big problem since the problems will still be caught when publishing an app, as they are today.

@eerhardt @agocke @michalstrehovsky @jtschuster

sbomer committed:
Annotate for trim/AOT (in)compatibility

sbomer committed:
Address AOT warnings not caught by analyzer

JamesNK commented on Src/Newtonsoft.Json/JsonTextReader.cs:
```diff
@@ -54,6 +56,8 @@ internal enum ReadType
     /// Represents a reader that provides fast, non-cached, forward-only access to JSON text data.
     /// </summary>
+    [RequiresUnreferencedCode(MiscellaneousUtils.TrimWarning)]
+    [RequiresDynamicCode(MiscellaneousUtils.AotWarning)]
```
Why add these annotations to types like JsonTextReader? I don't think it uses reflection or other problem APIs.

Same with reader, Newtonsoft.Json.Linq types, etc.

JamesNK replied:
I would like to tell people not to use anything to do with the serializer in AOT. But some Newtonsoft.Json things work and don't need warnings on them.

* Reader/writer should be fine in AOT.
* Almost all of JObject, JArray, JValue, etc should almost completely work. Those types have an `ToObject<T>()` method that invokes the serializer and should be annotated individually.

JamesNK replied:
Sorry, it's been a while since I've thought about trimming and AOT. I don't know what the feature switches mean. Maybe we should have a call sometime and you can talk through this PR and show me.

eerhardt commented on Src/Newtonsoft.Json/Newtonsoft.Json.csproj:
```diff
@@ -40,6 +40,8 @@
     Disable that warning since local builds do not strong-name sign the assembly, causing a mistmatch when comparing
     the previously shipped assemblies and current.-->
     <NoWarn Condition="'$(SignAssembly)' != 'true'">$(NoWarn);CP0003</NoWarn>
+    <IsTrimmable Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net6.0'))">true</IsTrimmable>
+    <IsAotCompatible Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net7.0'))">true</IsAotCompatible>
```
Thinking about this a bit deeper - we really only want the analyzers turned on. We don't want the `IsTrimmable` assembly metadata attribute to be put into the assembly. That is going to break existing apps that work today - like Blazor WASM.

```suggestion
    <EnableSingleFileAnalyzer Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net6.0'))">true</EnableSingleFileAnalyzer >
    <EnableTrimAnalyzer Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net6.0'))">true</EnableTrimAnalyzer >
    <EnableAotAnalyzerCondition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net7.0'))">true</EnableAotAnalyzer>
```

eerhardt replied:
> I would have expected trimming the serialization library not to make the problem much worse than it already is.

In a perfect world, this would be true. But I've seen crazier things. I would error on the side of not changing the behavior at all with this change.

eerhardt commented on Src/Newtonsoft.Json/Newtonsoft.Json.csproj:
```diff
@@ -51,6 +53,10 @@
   </ItemGroup>
+  <PropertyGroup Condition="'$(TargetFramework)'=='net8.0'">
+    <AssemblyTitle>Json.NET .NET 8.0</AssemblyTitle>
+    <DefineConstants>HAVE_ADO_NET;HAVE_APP_DOMAIN;HAVE_ASYNC;HAVE_ASYNC_DISPOSABLE;HAVE_BIG_INTEGER;HAVE_BINARY_FORMATTER;HAVE_BINARY_SERIALIZATION;HAVE_BINARY_EXCEPTION_SERIALIZATION;HAVE_CHAR_TO_LOWER_WITH_CULTURE;HAVE_CHAR_TO_STRING_WITH_CULTURE;HAVE_COM_ATTRIBUTES;HAVE_COMPONENT_MODEL;HAVE_CONCURRENT_COLLECTIONS;HAVE_COVARIANT_GENERICS;HAVE_DATA_CONTRACTS;HAVE_DATE_TIME_OFFSET;HAVE_DB_NULL_TY ...
```
This is a lot of duplication. Can we put this in a single spot and reuse it between `net8` and `net6`?

JamesNK replied:
Just keep the current standard. I'm not sure I like it anymore, but it would be annoying to change it.

sbomer committed:
Revert type-level annotations

sbomer committed:
Add more granular annotations

sbomer committed:
Remove annotations from XML nodes

sbomer committed:
Fix IsAotCompatible

Uncomment it and add a few missing fixes

sbomer committed:
Enable analyzers without IsTrimmable

eerhardt commented on Src/Newtonsoft.Json/Utilities/MiscellaneousUtils.cs:
```diff
@@ -40,6 +40,37 @@ namespace Newtonsoft.Json.Utilities
+        [FeatureSwitchDefinition("Newtonsoft.Json.SerializationIsSupported")]
+        [FeatureGuard(typeof(RequiresUnreferencedCodeAttribute))]
+        [FeatureGuard(typeof(RequiresDynamicCodeAttribute))]
+        internal static bool SerializationIsSupported => AppContext.TryGetSwitch("Newtonsoft.Json.SerializationIsSupported", out bool isSupported) ? isSupported : true;
```
My suggestion on these names is to name them like public .NET APIs. That way if we ever needed to expose this as a public property, we could match the signature to the name.

For example https://github.com/dotnet/runtime/blob/128fdfd1fd982e277e969d5c7ca1710345f4f649/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Helpers.cs#L27

eerhardt commented on Src/Newtonsoft.Json/Linq/JObject.cs:
```diff
@@ -828,7 +841,15 @@ EventDescriptorCollection ICustomTypeDescriptor.GetEvents()
         /// </returns>
         protected override DynamicMetaObject GetMetaObject(Expression parameter)
         {
+#if HAVE_APPCONTEXT
+            if (!MiscellaneousUtils.DynamicIsSupported)
```
Should this (and similar) be:

```suggestion
#if HAVE_DYNAMIC
            if (!MiscellaneousUtils.DynamicIsSupported)
```

?

eerhardt commented on Src/Newtonsoft.Json/Linq/JContainer.cs:
```diff
@@ -959,9 +959,17 @@ string ITypedList.GetListName(PropertyDescriptor[] listAccessors)
+#endif
             ICustomTypeDescriptor? d = First as ICustomTypeDescriptor;
 
+#pragma warning disable IL2026
```
I don't think these feature switches are going to work like you want them to. What you are doing is disabling the static analysis at build time, so users don't get warnings up front. But they will still get warnings at publish time. And their app will fail at runtime becuase no one defaults `ComponentModelIsSupported = false` in the build.

sbomer committed:
Make feature names match API shape

sbomer committed:
Remove old TODO

sbomer committed:
Fix exception messages

JamesNK commented on Src/Newtonsoft.Json/Utilities/TypeExtensions.cs:
```diff
@@ -593,27 +617,16 @@ public static bool AssignableToTypeName(this Type type, string fullTypeName, boo
                 current = current.BaseType();
             }
 
-            if (searchInterfaces)
-            {
-                foreach (Type i in type.GetInterfaces())
-                {
-                    if (string.Equals(i.Name, fullTypeName, StringComparison.Ordinal))
-                    {
-                        match = type;
-                        return true;
-                    }
-                }
-            }
```
Did you remove this because it caused trouble and `searchInterfaces` was always false?

JamesNK commented on Src/Newtonsoft.Json/Newtonsoft.Json.csproj:
```diff
@@ -40,6 +40,11 @@
     Disable that warning since local builds do not strong-name sign the assembly, causing a mistmatch when comparing
     the previously shipped assemblies and current.-->
     <NoWarn Condition="'$(SignAssembly)' != 'true'">$(NoWarn);CP0003</NoWarn>
+    <!-- Enable the analyzers directly instead of setting IsTrimmable or IsAotCompatible, to avoid adding IsTrimmable assembly metadata.
+         This should help avoid breaking consumers who use partial TrimMode. -->
+    <EnableSingleFileAnalyzer Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net6.0'))">true</EnableSingleFileAnalyzer>
+    <EnableTrimAnalyzer Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net6.0'))">true</EnableTrimAnalyzer>
+    <EnableAotAnalyzer Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net7.0'))">true</EnableAotAnalyzer>
```
Are these automatically enabled after .NET 8? It would be useful to have a comment about when they're neccessary so I know what to do when adding new targets.

JamesNK committed:
Improve JToken ToString and Write methods

JamesNK commented:
@sbomer I pushed some changes to improve JToken writing and ToString.

It looks like they're safe as long as there are no converters. I added new overloads without converters and used from ToString.

sbomer committed:
Annotate JsonSerializer at type level

sbomer committed:
Fix typos

sbomer committed:
Add missing AOT annotations

These were missed earlier because they weren't present
in the ref assembly:
https://github.com/dotnet/runtime/issues/95077

sbomer committed:
Add comment about TFMs for analyzer settings

JamesNK commented:
Looks good to me. Thanks!
